src/Microsoft.IdentityModel.Tokens.Saml/Saml2/Saml2SecurityTokenHandler.ValidateToken.Internal.cs

```diff
@@ -144,33 +154,21 @@ internal virtual ValidationResult<ValidatedConditions> ValidateConditions(
 
             if (samlToken.Assertion.Conditions.OneTimeUse)
             {
-                //ValidateOneTimeUseCondition(samlToken, validationParameters);
-                // We can keep an overridable method for this, or rely on the TokenReplayValidator delegate.
-                var oneTimeUseValidationResult = validationParameters.TokenReplayValidator(
-                    samlToken.Assertion.Conditions.NotOnOrAfter,
-                    samlToken.Assertion.CanonicalString,
-                    validationParameters,
-                    callContext);
+                var oneTimeUseValidationError = ValidateOneTimeUseCondition(samlToken, validationParameters, callContext);
 
-                if (!oneTimeUseValidationResult.IsValid)
-                {
-                    StackFrames.OneTimeUseValidationFailed ??= new StackFrame(true);
-                    return oneTimeUseValidationResult.UnwrapError().AddStackFrame(StackFrames.OneTimeUseValidationFailed);
-                }
+                if (oneTimeUseValidationError is not null)
+                    return oneTimeUseValidationError.AddCurrentStackFrame();
```

Comment:
Why do we return the error instead of throw it?

Reply:
This is the core of the work we are doing. Enhancing performance by removing exception throwing and returning validation errors that can be logged, thrown, or just checked on the above layers.

I think we have a deep dive coming on the topic, but basically by not throwing exceptions in some cases were are seeing over 30% performance improvement in our validation flows.